core/shared/src/main/scala/org/http4s/multipart/MultipartReceiver.scala

```diff
@@ -0,0 +1,192 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.multipart
+
+import cats.Applicative
+import cats.effect.Concurrent
+import cats.syntax.foldable._
+import fs2.io.file.Files
+import org.http4s.DecodeFailure
+import org.http4s.Headers
+import org.http4s.InvalidMessageBodyFailure
+import org.http4s.headers.`Content-Disposition`
+import org.typelevel.ci._
+
+trait MultipartReceiver[F[_], A] { self =>
+
+  /** Common supertype for the values decoded from each received part.
+    */
+  type Partial
+
+  /** Upon encountering the Headers for a new part, decide how (if at all)
+    * to handle the body of that part.
+    *
+    * Return  a `PartReceiver` wrapped in `Some` to use that receiver to
+    * decode the body of the part.
+    *
+    * Return `None` to indicate the part is unexpected. By default, this will
+    * result in a decode error upon encountering an unexpected part, but by
+    * calling [[ignoreUnexpectedParts]] you can create a MultipartReceiver that
+    * instead discards the body of any unexpected parts, without raising an error.
+    *
+    * @param partHeaders The `Headers` for a `Part` in a `multipart-form-data` request
+    * @return The receiver logic for that part
+    */
+  def decide(partHeaders: Headers): Option[PartReceiver[F, Partial]]
+
+  /** At the end of the decoding process, assemble a final result from the individual
+    * part-decode results that were returned by the `decide` method for each part.
+    *
+    * @param partials A list of decoded results as returned by the `decide` method for each part
+    * @return `Right` to indicate a successful result, or `Left` if something went wrong
+    */
+  def assemble(partials: List[Partial]): Either[DecodeFailure, A]
+
+  def decideOrReject(partHeaders: Headers): PartReceiver[F, Partial] =
+    decide(partHeaders).getOrElse {
+      PartReceiver.reject[F, Partial](MultipartReceiver.partName(partHeaders) match {
+        case None => InvalidMessageBodyFailure("Unexpected anonymous part")
+        case Some(name) => InvalidMessageBodyFailure(s"Unexpected part: '$name'")
+      })
+    }
+
+  def ignoreUnexpectedParts: MultipartReceiver[F, A] =
+    new MultipartReceiver.IgnoreUnexpected[F, A, Partial](this)
+
+  def transformParts(
+      f: PartReceiver[F, Partial] => PartReceiver[F, Partial]
+  ): MultipartReceiver[F, A] =
+    new MultipartReceiver.TransformPart[F, A, Partial](this, f)
+}
+
+object MultipartReceiver {
+  type Aux[F[_], A, P] = MultipartReceiver[F, A] { type Partial = P }
+
+  private def partName(headers: Headers) =
+    headers.get[`Content-Disposition`].flatMap(_.parameters.get(ci"name"))
+  private def partFilename(headers: Headers) =
+    headers.get[`Content-Disposition`].flatMap(_.parameters.get(ci"filename"))
```

Comment:
There are dragons in this header with internationalized names.  See #7419, #7436.  Does the receiver need some nicer way of interpreting these?

Reply:
Maybe I can just construct a `Part(headers, Stream.empty)` and delegate to its `name` and `filename` methods?